src/utils.rs

```diff
@@ -68,6 +72,79 @@ fn validate_keypair(pk: &str, sk: &str) -> ResultType<()> {
     Ok(())
 }
 
+fn doctor_tcp(address: std::net::IpAddr, port: &str, desc: &str) {
+    let start = std::time::Instant::now();
+    let conn = format!("{}:{}", address, port);
+    if let Ok(_stream) = TcpStream::connect(conn.as_str()) {
+        let elapsed = std::time::Instant::now().duration_since(start);
+        println!(
+            "TCP Port {} ({}): OK in {} ms",
+            port,
+            desc,
+            elapsed.as_millis()
+        );
+    } else {
+        println!("TCP Port {} ({}): ERROR", port, desc);
+    }
+}
+
+fn doctor_ip(server_ip_address: std::net::IpAddr, server_address: Option<&str>) {
+    println!("\nChecking IP address: {}", server_ip_address);
+    println!("Is IPV4: {}", server_ip_address.is_ipv4());
+    println!("Is IPV6: {}", server_ip_address.is_ipv6());
+
+    // reverse dns lookup
+    // TODO: (check) doesn't seem to do reverse lookup on OSX...
+    let reverse = lookup_addr(&server_ip_address).unwrap();
+    if server_address.is_some() {
+        if reverse == server_address.unwrap() {
+            println!("Reverse DNS lookup: '{}' MATCHES server address", reverse);
+        } else {
+            println!(
+                "Reverse DNS lookup: '{}' DOESN'T MATCH server address '{}'",
+                reverse,
+                server_address.unwrap()
+            );
+        }
+    }
+
+    // TODO: ICMP ping?
+
+    // port check TCP (UDP is hard to check)
+    doctor_tcp(server_ip_address, "21114", "API");
+    doctor_tcp(server_ip_address, "21115", "hbbs extra port for nat test");
+    doctor_tcp(server_ip_address, "21116", "hbbs");
+    doctor_tcp(server_ip_address, "21117", "hbbr tcp");
+    doctor_tcp(server_ip_address, "21118", "hbbs websocket");
+    doctor_tcp(server_ip_address, "21119", "hbbr websocket");
+
+    // TODO: key check
+}
+
+fn doctor(server_address_unclean: &str) {
+    let server_address3 = server_address_unclean.trim();
+    let server_address2 = server_address3.to_lowercase();
+    let server_address = server_address2.as_str();
+    println!("Checking server:  {}\n", server_address);
+    let server_ipaddr = server_address.parse::<IpAddr>();
+    if server_ipaddr.is_err() {
+        // the passed string is not an ip address
+        let ips: Vec<std::net::IpAddr> = lookup_host(server_address).unwrap();
+        println!("Found {} IP addresses: ", ips.iter().count());
+
+        for ip in ips.iter() {
+            println!(" - {}", ip);
+        }
```

Comment:
Nitpick: `ips.iter().inspect(|ip| println!(" - {ip}");`  is more "rustacean" 🤣

Reply:
it really looks better, but

```text
warning: unused `std::iter::Inspect` that must be used
   --> src/utils.rs:135:9
    |
135 |         ips.iter().inspect(|ip| println!(" - {ip}"));
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `#[warn(unused_must_use)]` on by default
    = note: iterators are lazy and do nothing unless consumed
```

Which is the better approach in this case?

```rust
        let _ = ips.iter().inspect(|ip| println!(" - {ip}"));
```

maybe this?

Reply:
`inspect` -> `for_each`

Reply:
fixed